### DISCUSSION TITLE
Is using <repository> references / <repository-reference> in category.xml / Tycho's <addIUTargetRepositoryReferences> considered best practice?

### DISCUSSION THREAD

**Discussion by sratz** [eclipse-equinox/p2#367] [General]
Using

- `<repository-reference>` in `category.xml`
- `<addIUTargetRepositoryReferences>` with Tycho (https://github.com/eclipse-tycho/tycho/blob/master/RELEASE_NOTES.md#new-option-to-filter-added-repository-references-when-assembling-a-p2-repository)

is a way to declare additional dependencies and not shipping them.

This causes p2 update sites to be added to user's installation, whenever a check-for-update operation is performed.

However, it also has some drawbacks:
1. This causes and excessive amount of update sites to accumulate in the 'Available Update Sites' preference page, e.g. this screenshot is from a standard Eclipse-for-Java installation after a single 'Check for update' which was canceled
![image](https://github.com/eclipse-equinox/p2/assets/14908423/65804ef0-25fe-453d-a0a1-a8a9239d2335)
Do we really want all those update sites with explicit versions to accumulate over time? There is no 'removal mechanism' when the 'parent' update site does not refer to them anymore.
2. It can cause issues where larger p2 update sites are implicitly added and things are available during check-for-update that are not actually intended. See https://github.com/eclipse-wildwebdeveloper/wildwebdeveloper/issues/938 and https://github.com/eclipse-wildwebdeveloper/wildwebdeveloper/issues/1380.
3. When using dynamic (`.../latest`) instead of fixed (`.../vX.Y`) update site URLs as dependencies, this can cause implicit, unintended updates of the IDE (also https://github.com/eclipse-wildwebdeveloper/wildwebdeveloper/issues/1380).

I feel like the usage of repository references has some caveats.

**Comment by HannesWell** [comment]
I would say that p2 should not add references from p2-repositories to the `Available Softare Sites`.
The references should only be considered while querying the repository but the references should not be presisted in a user Eclipse.
I'm surprised that this happens, but now that you said it, I also noticed that those references accumlate over time, but until know I assumed they are added explicitly by something like Oomph or as an explicit Update site for the product.

**Reply by merks** [reply]
It's possible that Oomph causes this problem, but I've not had time to check, though I think Oomph could/would only potentially cause such a problem if one performs a p2 task.   Of course touchpoints could add various things as well.

The above does not actually include instructions for how to reproduce a problem.  I would be good to have specific instructions such that I could try them on a "pure SDK" and also one with Oomph setup installed...

**Reply by laeubi** [reply]
Don't worry, I can confirm that this is a P2 feature that gave me some headache too.

To reproduce, simply install something form m2e site, or use a EPP, then remove all updatesites and add the ibuild repo to update to the latest ibuild results.

**Reply by laeubi** [reply]
And I think this is caused because the RepositoryManger.load is adding this to the "known update sites" of the profile and then it become a first class citizen in the update sites.

**Reply by sratz** [reply]
The initial trigger that caused me to look into this problem, is that we had been getting reports that customers' installations were updated to Java 21, even though this was not intended.

For the sake of completeness, here are the full steps to reproduce:
- Regular Java installation using Oomph and explicit selection of JustJ 17:
  ![image](https://github.com/eclipse-equinox/p2/assets/14908423/600aeb92-a2d5-42aa-9415-cc3572146fba)

- Update sites after first start:
  ![image](https://github.com/eclipse-equinox/p2/assets/14908423/c7efafeb-90e2-41db-a39a-2089cc44ef4e)

- Help -> Check for Updates:
  JustJ wants to be updated to 21, even though I explicitly chose JustJ 17 in the installer!
  ![image](https://github.com/eclipse-equinox/p2/assets/14908423/18233580-3821-4e94-9174-0ccc7f9bee32)

- Press 'Cancel'. Check 'Available Update Sites' again:
  ![image](https://github.com/eclipse-equinox/p2/assets/14908423/1a505186-95af-4cf9-9f09-ce44ffaf1dbb)

This is all but intuitive.

The reason why JustJ wants to be updated:
- https://download.eclipse.org/technology/m2e/releases/latest
  contains repository references
  - https://download.eclipse.org/eclipse/updates/4.28/
  - https://download.eclipse.org/lsp4e/releases/0.23.0/
  - https://download.eclipse.org/lsp4j/updates/releases/0.21.0/
  - https://download.eclipse.org/tm4e/releases/0.8.1/
  - https://download.eclipse.org/wildwebdeveloper/releases/1.3.0/
    contains repository references
    - https://download.eclipse.org/eclipse/updates/latest
    - https://download.eclipse.org/justj/jres/latest
    See bug https://github.com/eclipse-wildwebdeveloper/wildwebdeveloper/issues/1380

**Reply by mickaelistria** [reply]
I think m2e should repo not be referenced directly by EPP. project can setup different layouts that are not necessarily in the best interest of EPP (which wants to foresee everything included).

**Reply by merks** [reply]
It’s added by a touch point isn’t it?  It’s certainly not explicitly added by EPP.

**Reply by akurtakov** [reply]
I can definetely see both reasons why m2e latest should be included (to pick new m2e release/bugfixes even without simrel release) and also people wanting least amount of changes. And no matter how much I think  I can't imagine a way to satisfy both demands.
With the above said - this is already an issue with 2023-12 release , there hasn't been huge load of complains about getting on Java 21, 2024-06 will use Java 21 - should we actually change anything here? It's an update *proposed* but one has to explicitly apply it in order for the update to happen.

**Reply by laeubi** [reply]
> It’s added by a touch point isn’t it? It’s certainly not explicitly added by EPP.

It is [added by p2.inf](https://github.com/eclipse-m2e/m2e-core/blob/master/org.eclipse.m2e.feature/p2.inf) as other projects do as well. Beside that having any justj as a reference **is almost always wrong** in an _update-site_, but that's only a transitive dependency m2e is not controlling.

**Reply by akurtakov** [reply]
@laeubi Do you mean there is easier way to include jre? I haven't followed up with wwd lately but this is most likely coming from wwd shipping it's own product for which it wants the jre to be updated/kept current.

**Reply by laeubi** [reply]
You can add it to the products update sites for example:
![grafik](https://github.com/eclipse-equinox/p2/assets/1331477/52cb946d-dd3e-4826-a03e-60a78ef112ac)

**Reply by mickaelistria** [reply]
> You can add it to the products update sites for example

Anyone willing to change Wild Web Developer to adopt this practice is welcome to do so; we'll review/merge PRs promptly!

**Reply by merks** [reply]
I've prototyped the suggested change locally:

![image](https://github.com/eclipse-equinox/p2/assets/208716/84d82071-0458-4540-ad5c-cacbd25d5e5b)

Unfortunately it only "moves" the problem and arguably makes it even worse.

The plugin is configured to filter the IUs available from the referenced repositories:

![image](https://github.com/eclipse-equinox/p2/assets/208716/c062746c-bfbf-466f-a820-dec81ceaafd5)

So if I remove the reference, the repository directly contains JustJ JREs:

![image](https://github.com/eclipse-equinox/p2/assets/208716/cfcc9590-106c-4082-b144-f02fd3a2211c)

So this approach will not solve the problem of updating to Java 21 (or Java 22 when that is released) but rather makes it worse by locking in whatever version of JustJ was the latest at the time of the build such that one doesn't accidentally update to that latest Java 21 but rather to an older release of Java 21.

It would be nice if Tycho could do filtering without having to add a reference to the repository itself, i.e., perhaps specifying the repository URLs in the configuration directly, but that's currently not possible.   It's clear that what might be desired by WWD's product is not desirable for the EPP product. 

Keeping the product repository separate from the reuseable framework components' repository would be better for the consumers of WWD's components.  That is the approach I've taken for "my projects" that provide products:

https://ci.eclipse.org/cbi/view/p2RepoRelated/job/cbi.p2repo.aggregator-build/

**Reply by laeubi** [reply]
As said JustJ should *never* be part of the build / update site and then no need for filtering at all.

The JustJ site should only be added to the `eclipse-repository` project (e.g. using `<repository>` of layout `p2`) and then the *product* should contain the feature as a root (so it could be updated) and probably it can also add the JustJ site to the product so the user will get updates without explicitly adding JustJ ...

**Reply by merks** [reply]
It is already like this:

![image](https://github.com/eclipse-equinox/p2/assets/208716/66f986eb-d2f1-479d-9a7f-924a01887d15)

![image](https://github.com/eclipse-equinox/p2/assets/208716/c80094de-3352-4f1d-90d3-1ab5671feec3)

So I don't know how to configure things differently such that JustJ is not in the product repository and is not in a repository referenced by the product repository.  I think there no way, i.e., I think a product repository is always complete, except if filtered, and it can be fixed only by referencing a repository containing the thing to be filtered.  I might be wrong though and I would be happy to stand corrected.  I can try out whatever you might suggest...

**Reply by merks** [reply]
FYI, I did some debugging the other day when investigating if WWD could be tweaked.   The adding of repositories is done very deliberately by design where  org.eclipse.equinox.internal.p2.repository.helpers.AbstractRepositoryManager.remember(IRepository<T>, String) or org.eclipse.equinox.internal.p2.repository.helpers.AbstractRepositoryManager.remember(RepositoryInfo<T>, boolean) are called whenever some repository is added.  There is org.eclipse.equinox.internal.p2.metadata.repository.LocalMetadataRepository.publishRepositoryReferences() which sends out an event of type org.eclipse.equinox.internal.provisional.p2.repository.RepositoryEvent.DISCOVERED that is handled to add those repositories to the manager and are then remembered.  Because all this happens merely as a side-effect of loading, Oomph repository manager attempts to clean up those additions after loading:
```
        URI[] finalKnownRepositories = repositoryManager.getKnownRepositories(IRepositoryManager.REPOSITORIES_ALL);
        for (URI uri : finalKnownRepositories)
        {
          if (!originalKnownRepositories.contains(uri))
          {
            repositoryManager.removeRepository(uri);
          }
        }
```
But it's tricky to do this for p2 in general because installing something may involve touchpoints that add repositories such that one cannot simply restore the original state...

**Comment by laeubi** [comment]
> It would be nice if Tycho could do filtering without having to add a reference to the repository itself, i.e., perhaps specifying the repository URLs in the configuration directly, but that's currently not possible. 

Just want to put this on a different thread to not confuse things. The reason why Tycho works this way is because it is not a general filtering / modify repository action, this is supposed to remove otherwise required items (IUs+artifacts) and instead consume them from the original source! So having filtered but not adding a reference clearly makes no sense here, because then you will create a site that is missing items that a required and you can't install any software from that site anymore!

**Reply by merks** [reply]
Yes, I agree in general.  In this specific case though it's a product's repository and the product has this:

![image](https://github.com/eclipse-equinox/p2/assets/208716/82c2730d-d054-4d92-906b-00ed08ac93c9)

This was suggested as a way to solve the described problem and doesn't work because filtering does use this information.  But there is no point in belaboring the point...

**Reply by laeubi** [reply]
Product repositories are not filtered, so I'm a bit confused to be honest.

**Reply by merks** [reply]
Well, I'm not getting any less confused either.   😕 

The repository is specified to be filtered in the pom.xml and we can see in the published repository that the filtering has happened:

![image](https://github.com/eclipse-equinox/p2/assets/208716/0e304cc3-8504-4b3f-aa20-54eca0ef3d08)

This commit added the references (to the category.xml) and the filter specification (to the pom):

https://github.com/eclipse-wildwebdeveloper/wildwebdeveloper/commit/09f726978839807fdea667ca97303ebb231159ea

So we'd like a way to not have those reference in the repo and still have the filtering.  So I was merely suggesting that given the product is specify repositories that will be available during update, it might be possible to use those for filtering instead; of course that does not work currently and maybe that makes not as much sense.   Of course we'd also be happy to not have only a product repository but to have also a tools/framework repository, but I think anything we want is going involve self-service.  In any case, I think at this point it seems fair to conclude that we cannot achieve this goal with tools in their current form building only a product repository.  Again, I'd be happy to be wrong!

**Reply by laeubi** [reply]
> In any case, I think at this point it seems fair to conclude that we cannot achieve this goal with tools in their current form building only a product repository. Again, I'd be happy to be wrong!

Maybe it is because things are getting confused from udpate-sites versus "product repository", there is no such thing like a "product repository" in Tycho.

If we look [here](https://github.com/eclipse-wildwebdeveloper/wildwebdeveloper/tree/master/repository), the WWD defines a project with packaging `eclipse-repository` that contains 

- a [epp.product](https://github.com/eclipse-wildwebdeveloper/wildwebdeveloper/blob/master/repository/epp.product) 
- a [category.xml](https://github.com/eclipse-wildwebdeveloper/wildwebdeveloper/blob/master/repository/category.xml) 
- a [p2.inf](https://github.com/eclipse-wildwebdeveloper/wildwebdeveloper/blob/master/repository/p2.inf)

So what will happen here that first all content defined will be aggregated into a repository in the `target/repository` and from that it also contains an execution of the `materialize-products` mojo what will using this repository to "materialize" a zip/tar/... for all the configured platforms... and therefore **it must contain everything** either directly or indirectly and the result of this project is not necessarily suitable as an generic update-site it is only suitable as an update-site for the specific product(s)!

Instead one should have two modules/products (see [Structured layouts](https://tycho.eclipseprojects.io/doc/latest/StructuredBuild.html)), one "site" that contains the [category.xml](https://github.com/eclipse-wildwebdeveloper/wildwebdeveloper/blob/master/repository/category.xml)  and its result is be used as a generic update-site e.g. for consumers of WWD **as a library** (not referencing JustJ anywhere, possibly referencing referenced sites like used Eclipse Release **and is filtered**), and one "product" that contains the [epp.product](https://github.com/eclipse-wildwebdeveloper/wildwebdeveloper/blob/master/repository/epp.product)  (and possibly [p2.inf](https://github.com/eclipse-wildwebdeveloper/wildwebdeveloper/blob/master/repository/p2.inf) even though that is better placed in a feature or configured through the product) and just leave alone the `target/repository` (**what is not filtered**) but use the **produced materialized product** to deploy for download of the user. This product then will possibly reference the "site" URL+JustJ URL to allow update of the items inside the product.

That way consumers of WWD (e.g. m2e) will be prevented from "surprises" by getting product dependencies dripping in while the downloaded product can still be updating its features.

**Reply by merks** [reply]
This exactly confirms my understanding as well.   So the current design used by WWD is poor for consumers and can’t be fixed by tweaking the build.  The team has no plans to improve or change anything.

**Reply by laeubi** [reply]
> The team has no plans to improve or change anything.

But that's nothing that can be solved by Tycho/p2 or any configuration options :-)

**Comment by laeubi** [comment]
I have now addressed the problem with "surprising sites" here:
- https://github.com/eclipse-equinox/p2/pull/512

with that all described issues here will vanish because even though the JustJ is *referenced* it will not be used to search for updates (as long as the user not adding it explicitly).
